crates/ruff_linter/src/rules/pylint/rules/import_private_name.rs

```diff
@@ -0,0 +1,187 @@
+/// - [Semantic Versioning](https://semver.org/)
+///
+/// [PEP 8]: https://www.python.org/dev/peps/pep-0008/
+/// [PEP 420]: https://www.python.org/dev/peps/pep-0420/
+/// [`namespace-packages`]: https://beta.ruff.rs/docs/settings/#namespace-packages
+#[violation]
+pub struct ImportPrivateName {
+    name: String,
+    module: Option<String>,
+}
+
+impl Violation for ImportPrivateName {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let ImportPrivateName { name, module } = self;
+        match module {
+            Some(module) => {
+                format!("Private name import `{name}` from external module `{module}`")
+            }
+            None => format!("Private name import `{name}`"),
+        }
+    }
+}
+
+/// PLC2701
+pub(crate) fn import_private_name(
+    checker: &Checker,
+    scope: &Scope,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    for binding_id in scope.binding_ids() {
+        let binding = checker.semantic().binding(binding_id);
+        let Some(import) = binding.as_any_import() else {
+            continue;
+        };
+
+        let import_info = match import {
+            import if import.is_import() => ImportInfo::from(import.import().unwrap()),
+            import if import.is_from_import() => ImportInfo::from(import.from_import().unwrap()),
+            _ => return,
+        };
+
+        let Some(root_module) = import_info.module_name.first() else {
+            continue;
+        };
+
+        // Relative imports are not a public API.
+        // Ex) `from . import foo`
+        if import_info.module_name.starts_with(&["."]) {
+            continue;
+        }
+
+        // We can also ignore dunder names.
+        // Ex) `from __future__ import annotations`
+        // Ex) `from foo import __version__`
+        if root_module.starts_with("__") || import_info.member_name.starts_with("__") {
+            continue;
+        }
+
+        // Ignore private imports from the same module.
+        // Ex) `from foo import _bar` within `foo/baz.py`
+        if checker
+            .package()
+            .is_some_and(|path| path.ends_with(root_module))
+        {
+            continue;
+        }
+
+        // Ignore public imports; require at least one private name.
+        // Ex) `from foo import bar`
+        let Some((index, private_name)) = import_info
+            .call_path
+            .iter()
+            .find_position(|name| name.starts_with('_'))
+        else {
+            continue;
+        };
+
+        // Ignore private imports used exclusively for typing.
+        if !binding.references.is_empty()
+            && binding
+                .references()
+                .map(|reference_id| checker.semantic().reference(reference_id))
+                .all(is_typing)
+        {
+            continue;
+        }
+
+        let name = (*private_name).to_string();
+        let module = if index > 0 {
+            Some(import_info.module_name[..index].join("."))
```

Comment:
This doesn't seem to be safe. How is it guaranteed that `index` belongs to `module_name` and not the member name?

Reply:
Just my fault from a last-minute refactor prior to merging.

Reply:
#9553 slices over `call_path` instead.